README.md
## Redsmin Proxy — Securely connect behind the firewall (or localhost) Redis servers to [Redsmin](https://redsmin.com)

[![Build Status](https://img.shields.io/circleci/project/github/Redsmin/proxy.svg)](https://circleci.com/gh/Redsmin/proxy/) [![Coverage Status](https://img.shields.io/coveralls/Redsmin/proxy/master.svg)](https://coveralls.io/github/Redsmin/proxy?branch=master) [![Support](http://img.shields.io/badge/redsmin-support-0690fa.svg?style=flat)](https://redsmin.uservoice.com/) [![Follow](https://img.shields.io/twitter/follow/redsmin.svg?style=social)](https://twitter.com/Redsmin)

[![Docker Hub](http://img.shields.io/badge/docker-hub-22b8eb.svg?style=flat)](https://hub.docker.com/r/redsmin/proxy/) [![Docker hub](https://img.shields.io/docker/pulls/redsmin/proxy.svg)](https://hub.docker.com/r/redsmin/proxy/) [![NPM version](https://img.shields.io/npm/v/redsmin.svg)](http://badge.fury.io/js/redsmin) [![Downloads](http://img.shields.io/npm/dm/redsmin.svg)](https://www.npmjs.com/package/redsmin)


<p align="center">
<a target="_blank" href="https://redsmin.com"><img style="width:100%" src="docs/redsmin-proxy.png"></a>
</p>



- [Installation - Getting started](https://redsmin.uservoice.com/knowledgebase/articles/121169-can-i-manage-redis-instances-only-accessible-from-)
- [Environment variables options](#environment-variables-options)
- [Is the communication safe between my server and Redsmin? (Yes)](#is-the-communication-safe-between-my-server-and-redsmin-yes)
- How to start Redsmin proxy
  - [on Docker](#docker)
    -  [Start and connect Redsmin proxy to a local Redis server](#start-and-connect-redsmin-proxy-to-a-local-redis-server)
    -  [Start and connect Redsmin proxy to an network-wide available Redis](#start-and-connect-redsmin-proxy-to-an-network-wide-available-redis)
    -  [Start and connect Redsmin proxy to a Redis container](#start-and-connect-redsmin-proxy-to-a-redis-container)
    -  [Docker auto-restart](#docker-auto-restart)
  - [on Mac OS X](#macos-debianubuntu)
  - [on Debian/Ubuntu, *nix](#macos-debianubuntu)
  - [on Windows](#windows-powershell)
  - with a password protected redis
    - [Mac OS X](#StartWithPasswordProtectedRedisUnix)
    - [Debian/Ubuntu, *nix](#StartWithPasswordProtectedRedisUnix)
    - [Windows](#StartWithPasswordProtectedRedisWindows)
  - with a redis listening on a unix-socket
    - [Mac OS X](#how-to-start-redsmin-proxy-with-a-redis-listening-on-a-unix-socket)
    - [Debian/Ubuntu, *nix](#how-to-start-redsmin-proxy-with-a-redis-listening-on-a-unix-socket)
    - [Windows](#how-to-start-redsmin-proxy-with-a-redis-listening-on-a-unix-socket)
  - [reading a configuration file](#how-to-start-redsmin-proxy-reading-a-configuration-file)
- [How to connect multiple Redis from the same server to Redsmin](#macos-debianubuntu-4)
- How to keep redsmin proxy up once I disconnect
  - [with nohup](#with-nohup)
  - [with screen](#with-screen)
  - [with a process manager](#with-a-process-manager)
    - [with systemd](#with-systemd)
    - [with supervisord](#with-supervisord)
- [I'm behind a firewall, what rule should I add?](#im-behind-a-firewall-what-rule-should-i-add)
- [How to uninstall Redsmin Proxy](#how-to-uninstall-redsmin-proxy)
- [Throubleshooting](#throubleshooting)
  - [Ready check failed: NOAUTH Authentication required.](#ready-check-failed-noauth-authentication-required)
- [Changelog](/CHANGELOG.md)

We announce changes on our Twitter account [@redsmin](https://twitter.com/redsmin), our [Facebook page](https://www.facebook.com/redis.redsmin).

--------------------------------------------------------------------------------------------------

<a name="EnvironmentVariablesOptions"></a>
####  Environment variables options:  ####

- `CONFIG_FILE`: configuration file to read (if any), default: `/path/to/redsmin-proxy/etc/redsmin.json`
- `REDIS_URI`: Redis URI or socket path, default `redis://127.0.0.1:6379`
- `REDIS_AUTH`: Redis authentication password, default `null`
- `REDSMIN_KEY`: your Redsmin server connection key, default `''`

Advanced configuration:

- `REDSMIN_PORT`: where redsmin proxy should connect, default: `993`
- `REDSMIN_HOSTNAME`: where redsmin proxy should connect, default `ssl.redsmin.com`
- `DEBUG`: debug mode, default `false`
- Prefix `REDIS_URI` with `rediss://` to connect to Redis using TLS encryption

--------------------------------------------------------------------------------------------------

<a name="CommunicationSafety"></a>
#### Is the communication safe between my server and Redsmin? (Yes)

Yes, Redsmin and Redsmin proxy communicate through a secure connection using the [TLS 1.2 protocol](https://en.wikipedia.org/wiki/Transport_Layer_Security) so no one will be able to inspect the data looking at the traffic.

--------------------------------------------------------------------------------------------------


#### How to start Redsmin proxy

<a name="StartWithDocker"></a>
##### Docker


<a name="StartWithDockerAndConnectToALocalRedis"></a>
###### Start and connect Redsmin proxy to a local Redis server

Let say you started `redis-server` on your machine and then want to start redsmin-proxy from docker. **If you are on MacOSX or Windows the following command won't work** (if you are on Linux the following line will work):

```bash
docker run -it --rm --net=host --name redsmin-proxy -e REDSMIN_KEY=YOUR_REDSMIN_KEY -e REDIS_URI="redis://127.0.0.1:6379" redsmin/proxy
```

It does not work because on non-linux environment the docker daemon is running inside a VM and your `redis-server` is running on your host machine, thus accessing 127.0.0.1 from the docker daemon will simply hit the VM loopback.

So we simply need to specify the `HOST_IP` (replace it with your own local IP, you may want to use `ifconfig` to find it) instead of `127.0.0.1`:

```bash
docker run -it --rm --name redsmin-proxy -e REDSMIN_KEY=YOUR_REDSMIN_KEY -e REDIS_URI="redis://HOST_IP:6379" redsmin/proxy
```

On MacOSX, this should work and let redsmin-proxy connect to a Redis container on the same host:

```bash
docker run -it --rm --name redsmin-proxy -e REDSMIN_KEY=YOUR_REDSMIN_KEY -e REDIS_URI="redis://docker.for.mac.localhost:6379" redsmin/proxy
```

<a name="StartWithDockerAndConnectToNetworkWideRedis"></a>
###### Start and connect Redsmin proxy to an network-wide available Redis

```bash
docker run -it --rm --name redsmin-proxy -e REDSMIN_KEY=YOUR_REDSMIN_KEY -e REDIS_URI="redis://192.168.3.50:6379" redsmin/proxy
```

Where `redis://192.168.3.50:6379` will be the ip address and port of the running Redis server and `YOUR_REDSMIN_KEY` is your [Redsmin key](https://redsmin.uservoice.com/knowledgebase/articles/121169-can-i-manage-redis-instances-only-accessible-from).

<a name="StartWithDockerAndConnectToARedisContainer"></a>
###### Start and connect Redsmin proxy to a Redis container

Let first say you've started a Redis container:

```bash
docker run --name my-redis --rm redis
```

You can [link](https://docs.docker.com/userguide/dockerlinks/) redsmin proxy container to the redis one with `--link:

```bash
docker run -it --rm --name redsmin-proxy --link my-redis:local-redis -e REDSMIN_KEY=YOUR_KEY -e REDIS_URI="redis://local-redis:6379" redsmin/proxy
```

<a name="StartWithDockerAndUseAutoRestart"></a>
###### Docker auto-restart

If you want to leverage docker [auto-restart docker feature](https://docs.docker.com/reference/run/#restart-policies-restart), use the `--restart=always` command.


<a name="StartOnMacOSDebianUbuntu"></a>
##### MacOS, Debian/Ubuntu

```bash
npm install redsmin --global
REDIS_URI="redis://127.0.0.1:6379" REDSMIN_KEY="redsmin-token" redsmin
```

<a name="StartOnWindows"></a>
##### Windows (PowerShell)

```bash
npm install redsmin --global
$env:REDIS_URI="redis://127.0.0.1:6379"
$env:REDSMIN_KEY="redsmin-token"
redsmin
```

##### Windows (CMD.exe)

```bash
npm install redsmin --global
set REDIS_URI=redis://127.0.0.1:6379
set REDSMIN_KEY=redsmin-token
redsmin
```

Note: don't use double quotes for values on windows (bad 🔴 : `set KEY="VALUE"`, good ✅ : `set KEY=VALUE`)

--------------------------------------------------------------------------------------------------

<a name="StartWithPasswordProtectedRedis"></a>
#### How to start Redsmin proxy with a password protected redis

<a name="StartWithPasswordProtectedRedisUnix"></a>
##### MacOS, Debian/Ubuntu

```bash
npm install redsmin --global
REDIS_URI="redis://127.0.0.1:6379" REDIS_AUTH="auth-pass" REDSMIN_KEY="redsmin-token" redsmin
```

<a name="StartWithPasswordProtectedRedisWindows"></a>
##### Windows

```bash
npm install redsmin --global
set REDIS_URI=redis://127.0.0.1:6379
set REDIS_AUTH=auth-pass
set REDSMIN_KEY=redsmin-token
redsmin
```

Note: don't use double quotes for values on windows (bad 🔴 : `set KEY="VALUE"`, good ✅ : `set KEY=VALUE`)

--------------------------------------------------------------------------------------------------

<a name="StartWithRedisUnixSocket"></a>
#### How to start Redsmin proxy with a redis listening on a unix-socket

<a name="StartWithRedisUnixSocketUnix"></a>
##### MacOS, Debian/Ubuntu

```bash
npm install redsmin --global
REDIS_URI="/tmp/redis.sock" REDSMIN_KEY="5517e20046f4c7530d000357" redsmin
```

<a name="StartWithRedisUnixSocketWindows"></a>
##### Windows

```bash
npm install redsmin --global
set REDIS_URI=/tmp/redis.sock
set REDSMIN_KEY=5517e20046f4c7530d000357
redsmin
```

Note: 
- don't use double quotes for values on windows (bad 🔴 : `set KEY="VALUE"`, good ✅ : `set KEY=VALUE`)
- you may need to use `sudo` to access to the socket.

--------------------------------------------------------------------------------------------------

<a name="startWithConfigurationFile"></a>
#### How to start Redsmin proxy reading a configuration file

First create a json configuration file, for instance  `/etc/redsmin.json`:

```json
{
  "key": "redsmin-token",
  "redis": "redis://127.0.0.1:6379",
  "auth": ""
}
```

Then start redsmin proxy with:

##### MacOS, Debian/Ubuntu

```bash
CONFIG_FILE="/etc/redsmin.json" redsmin
```

##### Windows

```bash
set CONFIG_FILE="/etc/redsmin.json"
redsmin
```

--------------------------------------------------------------------------------------------------

<a name="connectMultipleRedis">
  
#### How to connect multiple Redis from the same server to Redsmin

##### MacOS, Debian/Ubuntu

```bash
REDIS_URI="redis://127.0.0.1:6379" REDSMIN_KEY="redsmin-token1" redsmin &
REDIS_URI="redis://127.0.0.1:6380" REDSMIN_KEY="redsmin-token2" redsmin &
REDIS_URI="redis://127.0.0.1:6381" REDSMIN_KEY="redsmin-token3" redsmin &
REDIS_URI="redis://127.0.0.1:6382" REDSMIN_KEY="redsmin-token4" redsmin &
```

##### Windows

```bash
set REDIS_URI="redis://127.0.0.1:6379"
set REDSMIN_KEY=redsmin-token1
START /B redsmin

set REDIS_URI="redis://127.0.0.1:6380"
set REDSMIN_KEY=redsmin-token2
START /B redsmin

set REDIS_URI="redis://127.0.0.1:6381"
set REDSMIN_KEY=redsmin-token3
START /B redsmin

set REDIS_URI="redis://127.0.0.1:6382"
set REDSMIN_KEY=redsmin-token4
START /B redsmin
```

Note: 
- of course we could have used multiple `CONFIG_FILE` instead of environment variables.

--------------------------------------------------------------------------------------------------

#### How to keep redsmin proxy up once I disconnect

<a name="keepRedsminProxyUpWithNohup"></a>
## With nohup

The easiest way is to use [nohup](http://linux.die.net/man/1/nohup) that will keep redsmin-proxy running even once the SSH session is closed. Simply connect to the server that contains Redis, run the commands below, don't forget to replace `YOUR_REDSMIN_TOKEN` with the `REDSMIN_TOKEN` you had when creating the proxy connection from Redsmin app.

```shell
echo '#!/usr/bin/env bash' >> redsmin-proxy.sh
echo 'while true; do REDSMIN_KEY=YOUR_REDSMIN_TOKEN redsmin; sleep 1; done;' >> redsmin-proxy.sh
chmod +x redsmin-proxy.sh
nohup ./redsmin-proxy.sh &
```

To check that everything is alright or to debug Redsmin proxy, you can use `tail -f nohup.out`.

## With nohup (one-liner)

```
nohup bash -c "while true; do REDSMIN_KEY=YOUR_REDSMIN_TOKEN redsmin; sleep 1; done" &
```

<a name="keepRedsminProxyUpWithScreen"></a>
## With screen

On MacOS, Ubuntu/Debian, the simplest way is to use [screen](http://www.rackaid.com/blog/linux-screen-tutorial-and-how-to/):

```
# start screen
screen
# start redsmin-proxy
REDIS_URI="redis://127.0.0.1:6379" REDSMIN_KEY="redsmin-token1" redsmin
# Ctrl+A+D to detach from screen
# and then to reattach to the screen session:
screen -r
```

<a name="keepRedsminProxyUpWithAProcessManager"></a>
## With a process manager

But you could also use [Upstart](http://upstart.ubuntu.com/), [systemd](http://www.freedesktop.org/wiki/Software/systemd/), [supervisord](http://supervisord.org/) or [pm2](https://github.com/Unitech/PM2) on these system.

On Windows you will need to [create a service](http://support.microsoft.com/en-us/kb/251192) or use [pm2](https://github.com/Unitech/PM2).

## With Systemd

Create the service at `/etc/systemd/system/redsmin.service`

```
[Unit]
Description = Redsmin Proxy
After = network.target

[Service]
Type = simple
Environment = REDIS_URI=redis://127.0.0.1:6379 REDSMIN_KEY=your-token-here
ExecStart = /usr/bin/redsmin $REDIS_URI $REDSMIN_KEY
TimeoutStartSec = infinity
Restart = on-abort

[Install]
WantedBy = multi-user.target
```

Reload systemd by `systemctl daemon-reload`.

You can now start/stop/restart redsmin like any other systemd service, like `systemctl start redsmin`.

## With Supervisord

Create a config file with these contents:

```
[program:redsmin]
command = /usr/bin/redsmin
autostart = true
autorestart = true
environment = REDIS_URI="redis://127.0.0.1:6379",REDSMIN_KEY="your-token-here"
```

Reload supervisord config by `supervisorctl reread && supervisorctl update`

**We will happily merge into this repository any pull-request describing a configuration file for any other process runner.**

--------------------------------------------------------------------------------------------------

<a name="RuleBehindTheFirewall"></a>
#### I'm behind a firewall, what rule should I add?

Redsmin proxy connects to `ssl.redsmin.com` on port `993` with a secure [TLS socket connection](https://en.wikipedia.org/wiki/Transport_Layer_Security). For troubleshooting: [What ip/port should I locally open to use Redsmin proxy](https://redsmin.uservoice.com/knowledgebase/articles/274294-what-ip-port-should-i-locally-open-to-use-redsmin-).

--------------------------------------------------------------------------------------------------

<a name="Uninstall"></a>
#### How to uninstall Redsmin Proxy

##### MacOS, Debian/Ubuntu

```bash
npm uninstall redsmin -g
```

<a name="troubleshooting"></a>
#### Throubleshooting

<a name="noauth"></a>
##### Ready check failed: NOAUTH Authentication required

It means that your Redis server required a password and that no password is configured in Redsmin Proxy. To fix this start Redsmin proxy with the [`REDIS_AUTH`](#environment-variables-options) environment variable.


CHANGELOG.md
## <small>5.0.1 (2020-05-08)</small>

* Release v5.0.1. ([e8f793a](https://github.com/Redsmin/proxy/commit/e8f793a))
* fix: logger ([700418a](https://github.com/Redsmin/proxy/commit/700418a))
* docs(changelog): updated ([f7afdec](https://github.com/Redsmin/proxy/commit/f7afdec))



## 5.0.0 (2020-05-08)

* Release v5.0.0. ([018532c](https://github.com/Redsmin/proxy/commit/018532c))



## <small>4.1.4 (2020-05-08)</small>

* Release v4.1.4. ([7d7cbb1](https://github.com/Redsmin/proxy/commit/7d7cbb1))
* fix: docker image ([f62661c](https://github.com/Redsmin/proxy/commit/f62661c))
* docs(changelog): updated ([597b449](https://github.com/Redsmin/proxy/commit/597b449))



## <small>4.1.3 (2020-05-08)</small>

* Release v4.1.3. ([3716b53](https://github.com/Redsmin/proxy/commit/3716b53))
* docs(changelog): updated ([9aaab57](https://github.com/Redsmin/proxy/commit/9aaab57))
* docs(changelog): updated ([364e49c](https://github.com/Redsmin/proxy/commit/364e49c))
* fix: package ([4d7dc0f](https://github.com/Redsmin/proxy/commit/4d7dc0f))
* fix: tests ([6c5704e](https://github.com/Redsmin/proxy/commit/6c5704e))
* chore(deps-dev): bump mocha from 3.5.0 to 3.5.3 ([c47dbbc](https://github.com/Redsmin/proxy/commit/c47dbbc))
* chore(deps): [security] bump extend from 3.0.0 to 3.0.2 ([1bf47ce](https://github.com/Redsmin/proxy/commit/1bf47ce))
* chore(deps): [security] bump hawk from 3.1.0 to 3.1.3 ([dd71b25](https://github.com/Redsmin/proxy/commit/dd71b25))
* chore(deps): [security] bump is-my-json-valid from 2.12.2 to 2.20.0 ([e0712db](https://github.com/Redsmin/proxy/commit/e0712db))
* chore(deps): bump common-env from 6.3.0 to 6.3.1 ([d14e813](https://github.com/Redsmin/proxy/commit/d14e813))



## <small>4.1.2 (2020-03-12)</small>

* fix ([e6c3318](https://github.com/Redsmin/proxy/commit/e6c3318))
* fix ([4d4bc21](https://github.com/Redsmin/proxy/commit/4d4bc21))
* Release v4.1.1. ([d3674e6](https://github.com/Redsmin/proxy/commit/d3674e6))
* Release v4.1.2. ([5107644](https://github.com/Redsmin/proxy/commit/5107644))
* Update README.md ([683dbe0](https://github.com/Redsmin/proxy/commit/683dbe0))
* Update README.md ([9749e65](https://github.com/Redsmin/proxy/commit/9749e65))
* Update README.md ([3e56c5e](https://github.com/Redsmin/proxy/commit/3e56c5e))
* fix: pkg lock ([e89a0d1](https://github.com/Redsmin/proxy/commit/e89a0d1))
* fix(package): upgrade deps ([6ae310f](https://github.com/Redsmin/proxy/commit/6ae310f))
* chore(deps): [security] bump bl from 1.0.0 to 1.0.3 ([5b1202a](https://github.com/Redsmin/proxy/commit/5b1202a))
* chore(deps): [security] bump brace-expansion from 1.1.1 to 1.1.11 ([5c868d6](https://github.com/Redsmin/proxy/commit/5c868d6))
* chore(deps): [security] bump concat-stream from 1.5.1 to 1.6.2 ([60c4371](https://github.com/Redsmin/proxy/commit/60c4371))
* chore(deps): [security] bump extend from 3.0.0 to 3.0.2 ([689c49a](https://github.com/Redsmin/proxy/commit/689c49a))
* chore(deps): [security] bump fstream from 1.0.10 to 1.0.12 ([1726af0](https://github.com/Redsmin/proxy/commit/1726af0))
* chore(deps): [security] bump handlebars from 4.0.5 to 4.5.3 ([ce0837b](https://github.com/Redsmin/proxy/commit/ce0837b))
* chore(deps): [security] bump hawk from 3.1.0 to 3.1.3 ([2ef0e6d](https://github.com/Redsmin/proxy/commit/2ef0e6d))
* chore(deps): [security] bump is-my-json-valid from 2.12.2 to 2.20.0 ([fb8677d](https://github.com/Redsmin/proxy/commit/fb8677d))
* chore(deps): [security] bump lodash from 4.17.4 to 4.17.13 ([bd3376f](https://github.com/Redsmin/proxy/commit/bd3376f))
* chore(deps): [security] bump lodash.template from 4.4.0 to 4.5.0 ([f74bdf8](https://github.com/Redsmin/proxy/commit/f74bdf8))
* chore(deps): [security] bump moment from 2.10.6 to 2.24.0 ([3ce80ac](https://github.com/Redsmin/proxy/commit/3ce80ac))
* chore(deps): [security] bump node-uuid from 1.4.3 to 1.4.8 ([133aa9d](https://github.com/Redsmin/proxy/commit/133aa9d))
* chore(deps): [security] bump request from 2.65.0 to 2.74.0 ([dd2d7c9](https://github.com/Redsmin/proxy/commit/dd2d7c9))
* chore(deps): [security] bump sshpk from 1.9.2 to 1.16.1 ([a1840a8](https://github.com/Redsmin/proxy/commit/a1840a8))
* chore(deps): [security] bump tar from 2.2.1 to 2.2.2 ([1b3d25a](https://github.com/Redsmin/proxy/commit/1b3d25a))
* chore(deps): [security] bump tough-cookie from 2.2.0 to 2.3.4 ([94b345e](https://github.com/Redsmin/proxy/commit/94b345e))
* chore(deps): bump common-env from 6.1.0 to 6.3.0 ([e6432e9](https://github.com/Redsmin/proxy/commit/e6432e9))
* chore(deps): bump debug from 4.0.1 to 4.1.1 ([5a0a0c9](https://github.com/Redsmin/proxy/commit/5a0a0c9))
* chore(deps): bump lodash from 4.17.13 to 4.17.15 ([c41b8d7](https://github.com/Redsmin/proxy/commit/c41b8d7))
* doc: updated README, removed double quotes for set env on windows ([5f652d5](https://github.com/Redsmin/proxy/commit/5f652d5))
* docs(changelog): updated ([011849c](https://github.com/Redsmin/proxy/commit/011849c))



## 4.1.0 (2018-10-30)

* Release v4.1.0. ([1f112a5](https://github.com/Redsmin/proxy/commit/1f112a5))



## <small>4.0.2 (2018-10-30)</small>

* Add Systemd script example ([e85eee4](https://github.com/Redsmin/proxy/commit/e85eee4))
* Add test for TLS connect ([95d09df](https://github.com/Redsmin/proxy/commit/95d09df))
* feat(gitlab) ([968d2c4](https://github.com/Redsmin/proxy/commit/968d2c4))
* fix(test) ([7a80f5d](https://github.com/Redsmin/proxy/commit/7a80f5d))
* fix(test) ([d4dc333](https://github.com/Redsmin/proxy/commit/d4dc333))
* Release v4.0.2. ([0f3bfee](https://github.com/Redsmin/proxy/commit/0f3bfee))
* Reove PID ([5fa1e42](https://github.com/Redsmin/proxy/commit/5fa1e42))
* Support TLS connection to Redis ([bb081c9](https://github.com/Redsmin/proxy/commit/bb081c9))
* Update can_handshake.sh ([593c4be](https://github.com/Redsmin/proxy/commit/593c4be))
* Update can_handshake.sh ([3487f1f](https://github.com/Redsmin/proxy/commit/3487f1f))

lib/Proxy.test.js
'use strict';
var _ = require('lodash');
var sinon = require('sinon');
var tls = require('tls');
var fs = require('fs');
var path = require('path');
var net = require('net');
var log = require('./log')('debug');
var jsonPackage = require('../package.json');
var ProxyFactory = require('./Proxy');
var RedisClient = require('./RedisClient')(log, net, tls);
var Endpoint = require('./Endpoint')(log, jsonPackage, tls, process);
var debug = require('debug')('Proxy.test');

process.env.NODE_TLS_REJECT_UNAUTHORIZED = "0";
describe('Proxy', function () {
  var address;
  var redsminServer;

  beforeEach(function(done) {

    redsminServer = tls.createServer({
      key: fs.readFileSync(path.resolve(__dirname, 'server-key.fixtures.pem')),
      cert: fs.readFileSync(path.resolve(__dirname, 'server-cert.fixtures.pem')),
    }, function (socket) {
      socket.on('data', function(data){
        console.log('got data', data);
      });
      debug('got a new socket');
    }).on('error', function(err) {
      // handle errors here
      throw err;
    });

    redsminServer.listen(function(){
      address = redsminServer.address();
      debug('opened redsminServer on %j', address);

      done();
    });
  });

  afterEach(function(done) {
    redsminServer.close(done);
  });

  it('should start', function () {
    var config = {
      debug: true,
      redsmin: {
        port: address.port,
        hostname: '127.0.0.1',
        key: 'REDSMIN_KEY',
        auth: 'auth'
      },
      redis:{
        auth: 'plop',
        uri: '127.0.0.1:10'
      }
    };

    var Proxy = ProxyFactory(config, RedisClient, Endpoint);
    var proxy = new Proxy();
    proxy.start();

    setTimeout(function(){
      proxy._stop();
    }, 1000);
  });
});


lib/config.test.js
'use strict';
var configFactory = require('./config');
var t = require('chai').assert;
var fs = require('fs');
var path = require('path');

describe('config', function () {
  var Config, config;

  beforeEach(function (done) {
    delete process.env.REDIS_URI;
    delete process.env.REDIS_AUTH;
    Config = configFactory(console);
    config = new Config(path.resolve(__dirname, '../test/fixtures/redsmin_default.json'));
    done();
  });

  it('should return a configuration', function () {
    t.strictEqual(config.redis.auth, null);
    t.strictEqual(config.redis.uri, 'redis://127.0.0.1:6379');
    t.strictEqual(config.redsmin.key, '5331e06500617e0b090000aa');
    t.strictEqual(config.redsmin.port, 993);
    t.strictEqual(config.redsmin.hostname, 'ssl.redsmin.com');
  });

  it('should use env vars if defined', function () {
    process.env.REDIS_URI = '/tmp/redis.sock';
    process.env.REDIS_AUTH = 'aaaa';
    config = new Config(path.resolve(__dirname, '../test/fixtures/redsmin_default.json'));
    t.strictEqual(config.redis.uri, process.env.REDIS_URI);
    t.strictEqual(config.redis.auth, process.env.REDIS_AUTH);
  });
});


test/Proxy.test.js
var Proxy = require('../lib/Proxy')
,   _        = require('lodash')
,   sinon    = require('sinon');

// Quiet console output
Proxy.log         = sinon.stub(_.clone(console));

exports['Proxy'] = {
  setUp: function(done) {
    done();
  },

  tearDown: function (callback) {
    // clean up
    //Endpoint.log = console;
    callback();
  },

  'constructor': function(t) {
    t.done();
  }
};


app.js
'use strict';

/**
 * Redsmin proxy daemon
 * Francois-Guillaume Ribreau - @FGRibreau
 * redsmin.com
 */

var fs = require('fs');
var net = require('net');
var tls = require('tls');
var path = require('path');
var updateNotifier = require('update-notifier');
var pkg = require('./package.json');

var log = require('./lib/log')('debug');
var Config = require('./lib/config')(log);
var config = new Config(Config.DEFAULT_CONFIG_PATH);

var jsonPackage = JSON.parse(fs.readFileSync(path.resolve(__dirname, './package.json')));

var Endpoint = require('./lib/Endpoint')(log, jsonPackage, tls, process);

var RedisClient = require('./lib/RedisClient')(log, net, tls);

var RedsminProxy = require('./lib/Proxy')(config, RedisClient, Endpoint);

updateNotifier({
  pkg:pkg,
  updateCheckInterval: 1000*60 //  1 minute
}).notify();

// Start the proxy
new RedsminProxy().start();


grunt.js
module.exports = function(grunt) {
  // Project configuration
  grunt.initConfig({
    pkg: '<json:package.json>',

    lint: {
      all: ['*.js', 'lib/**/*.js', 'test/**/*.js']
    },

    jshint: {
      options: {
        curly: true,
        es5: true,
        eqeqeq: true,
        immed: true,
        latedef: true,
        laxcomma: true,
        newcap: false,
        noarg: true,
        sub: true,
        undef: true,
        eqnull: true,
        browser: false
      },
      globals: {
        module: true,
        require:true,
        exports:true,
        console:true,
        __dirname:true,
        process:true
      }
    },

    watch: {
      files: ['<config:lint.all>'],
      tasks: 'lint test'
    },

    test: {
      all: ['test/**/*.js']
    }
  });

  // http://blog.fgribreau.com/2012/06/how-to-get-growl-notifications-from.html
  var growl = require('growl');
  ['warn', 'fatal'].forEach(function(level) {
    grunt.utils.hooker.hook(grunt.fail, level, function(opt) {
      growl(opt.name, {
        title: opt.message,
        image: 'Console'
      });
    });
  });

  // Default task.
  grunt.registerTask('default', 'lint test');
};


Gruntfile.js
module.exports = function (grunt) {
  // Project configuration
  grunt.initConfig({
    pkg: '<json:package.json>',

    lint: {
      all: ['*.js', 'lib/**/*.js', 'test/**/*.js']
    },

    jshint: {
      options: {
        curly: true,
        es5: true,
        eqeqeq: true,
        immed: true,
        latedef: true,
        laxcomma: true,
        newcap: false,
        noarg: true,
        sub: true,
        undef: true,
        eqnull: true,
        browser: false
      },
      globals: {
        module: true,
        require: true,
        exports: true,
        console: true,
        __dirname: true,
        process: true
      }
    },

    watch: {
      files: ['<config:lint.all>'],
      tasks: 'lint test'
    },

    test: {
      all: ['test/**/*.js']
    }
  });

  // Default task.
  grunt.registerTask('default', 'lint test');
};


lib/RedisClient.js
'use strict';

var _ = require('lodash');
var url = require('url');
var backoff = require('backoff');
var EventEmitter = require('events').EventEmitter;
var assert = require('assert');

module.exports = function (log, net, tls) {
  assert(_.isObject(log));
  assert(_.isObject(net));


  /**
   * @class Redis client
   * @param {Function} fnWrite `callback(data)` where to write datas from redis
   */
  function RedisClient(fnWrite, endpoint) {
    EventEmitter.call(this);
    _.bindAll(this, _.functionsIn(this));

    if (!fnWrite || typeof fnWrite !== 'function') {
      throw new Error("RedisClient `fnWrite` parameter is not defined or is not a function");
    }

    this.connected = false;
    this.endpoint = endpoint;

    /**
     * Endpoint instance
     * @type {Endpoint}
     */
    this.fnWrite = fnWrite;

    this.port = null;
    this.hostname = null;
    this.socket = null;

    this.backoff = backoff.fibonacci({
      initialTimeout: 1,
      maxTimeout: 1000
    });

    this.backoff.on('backoff', this.reconnect);
  }

  _.extend(RedisClient.prototype, require('events').EventEmitter.prototype, {
    PREFIX: 'redis://',
    TLS_PREFIX: 'rediss://',

    updatePortAndHostname: function (uri) {
      uri = uri || this.PREFIX + '127.0.0.1:6379';

      if (uri[0] !== '/' && uri.indexOf(this.PREFIX) !== 0 && uri.indexOf(this.TLS_PREFIX) !== 0) { // add "redis://" if it was not specified
        uri = this.PREFIX + uri;
      }

      if (uri[0] !== '/') { // not a socket path
        var parsedUri = url.parse(uri);
        this.uri = uri;
        this.port = parsedUri.port ||  6379;
        this.hostname = parsedUri.hostname || '127.0.0.1';
        this.isTLS = (parsedUri.protocol + '//' === this.TLS_PREFIX);
      } else {
        this.uri = uri; // a socket path
        this.port = null;
        this.hostname = null;
        this.isTLS = false;
      }
    },

    connect: function (uri) {
      this.updatePortAndHostname(uri);
      this._connect();
    },

    _connect: function () {
      if (this.socket) {
        this.socket.removeAllListeners();
        this.socket.destroy();
      }

      if (this.port && this.hostname) {
        // port
        this.socket = this.isTLS ? tls.connect(this.port, this.hostname, this.onConnected) : net.createConnection(this.port, this.hostname, this.onConnected);
      } else {
        // socket
        this.socket = net.createConnection({
          path: this.uri
        }, this.onConnected);
      }

      /**
       * Enable/disable keep-alive functionality, and optionally set the initial delay
       * before the first keepalive probe is sent on an idle socket. enable defaults to false.
       */
      this.socket.setKeepAlive(true);

      /**
       * Disables the Nagle algorithm. By default TCP connections use the Nagle algorithm,
       * they buffer data before sending it off. Setting true for noDelay will immediately
       * fire off data each time socket.write() is called. noDelay defaults to true.
       */
      this.socket.setNoDelay(true);

      this.socket.setTimeout(0);

      this.socket.on('data', this.onData);
      this.socket.on('close', this.onClose);
      this.socket.on('error', this.onError);
    },

    reconnect: function ( /*number , delay*/ ) {
      if (this.connected) {
        return this.backoff.reset();
      }

      log.info("[RedisClient] Reconnecting...");
      this._connect();
    },

    onConnected: function () {
      this.connected = true;
      this.backoff.reset();
      log.info("[RedisClient] Redis client connected to " + this.uri);
      this.emit('connect');
    },

    /**
     * Forward data from Redis to the fnWrite
     * @param  {String} data data from redis
     */
    onData: function (data) {
      this.fnWrite(data);
    },

    /**
     * Forward data from elsewhere to Redis
     */
    write: function (data) {
      this.socket.write(data);
    },

    /**
     * If the connection to redis just closed, try to reconnect
     * @param  {Error} err
     */
    onClose: function (err) {
      log.error("Redis client closed " + (err ? err.message : ''));
      this.connected = false;
      this.emit('close', err);
      this.backoff.reset();
      this.backoff.backoff();
    },

    onError: function (err) {
      if (!err) {
        return;
      }

      log.error('Redis client error ' + (err ? err.message : ''));
    }
  });

  return RedisClient;
};
